pcap-dpdk.c

```diff
@@ -0,0 +1,740 @@
+#include <rte_bus.h>
+
+#include "pcap-int.h"
+#include "pcap-dpdk.h"
+
+#define DPDK_LIB_NAME "libpcap_dpdk"
+#define DPDK_ARGC_MAX 64 
+#define DPDK_CFG_MAX_LEN 1024
+#define DPDK_CFG_ENV_NAME "DPDK_CFG"
+static char dpdk_cfg_buf[DPDK_CFG_MAX_LEN];
+#define DPDK_PCI_ADDR_SIZE 16
+#define DPDK_DEF_CFG "--log-level=debug -l0 -dlibrte_pmd_e1000.so -dlibrte_pmd_ixgbe.so -dlibrte_mempool_ring.so"
+#define DPDK_PREFIX "dpdk:"
+#define MBUF_POOL_NAME "mbuf_pool"
+#define DPDK_TX_BUF_NAME "tx_buffer"
+//The number of elements in the mbuf pool.
+#define DPDK_NB_MBUFS 8192U
+#define MEMPOOL_CACHE_SIZE 256
+#define MAX_PKT_BURST 32 
+// Configurable number of RX/TX ring descriptors
+#define RTE_TEST_RX_DESC_DEFAULT 1024
+#define RTE_TEST_TX_DESC_DEFAULT 1024
+
+static uint16_t nb_rxd = RTE_TEST_RX_DESC_DEFAULT;
+static uint16_t nb_txd = RTE_TEST_TX_DESC_DEFAULT;
+
+#define RTE_ETH_PCAP_SNAPLEN ETHER_MAX_JUMBO_FRAME_LEN
+
+static struct rte_eth_dev_tx_buffer *tx_buffer;
+
+struct dpdk_ts_helper{
+	struct timeval start_time;
+	uint64_t start_cycles;
+	uint64_t hz;
+};
+struct pcap_dpdk{
+	pcap_t * orig;
+	uint16_t portid; // portid of DPDK
+	pcap_handler cb; //callback and argument
+	u_char *cb_arg;
+	int max_cnt;
+	int must_clear_promisc;
+	int filter_in_userland;
+	uint64_t rx_pkts;
+	uint64_t bpf_drop;
+	struct ether_addr eth_addr;
+	struct timeval prev_ts;
+	struct rte_eth_stats prev_stats;
+	struct timeval curr_ts;
+	struct rte_eth_stats curr_stats;
+	uint64_t pps;
+	uint64_t bps;
+	struct rte_mempool * pktmbuf_pool;
+	struct dpdk_ts_helper ts_helper;
+	char pci_addr[DPDK_PCI_ADDR_SIZE];
+	unsigned char pcap_tmp_buf[RTE_ETH_PCAP_SNAPLEN];
+	volatile sig_atomic_t break_loop;
+};
+
+static struct rte_eth_conf port_conf = {
+	.rxmode = {
+		.split_hdr_size = 0,
+	},
+	.txmode = {
+		.mq_mode = ETH_MQ_TX_NONE,
+	},
+};
+
+static int dpdk_init_timer(struct pcap_dpdk *pd){
+	gettimeofday(&(pd->ts_helper.start_time),NULL);
+	pd->ts_helper.start_cycles = rte_get_timer_cycles();
+	pd->ts_helper.hz = rte_get_timer_hz();
+	if (pd->ts_helper.hz == 0){
+		return -1;	
+	}
+	return 0;
+}
+static inline void calculate_timestamp(struct dpdk_ts_helper *helper,struct timeval *ts)
+{
+	uint64_t cycles;
+	// delta
+	struct timeval cur_time;
+	cycles = rte_get_timer_cycles() - helper->start_cycles;
+	cur_time.tv_sec = (time_t)(cycles/helper->hz);
+	cur_time.tv_usec = (suseconds_t)((cycles%helper->hz)*1e6/helper->hz);
+	timeradd(&(helper->start_time), &cur_time, ts);
+}
+
+static unsigned int dpdk_gather_data(unsigned char *data, struct rte_mbuf *mbuf)
+{
+	unsigned int total_len = 0;
+	while (mbuf && (total_len+mbuf->data_len) < RTE_ETH_PCAP_SNAPLEN ){
+		rte_memcpy(data+total_len, rte_pktmbuf_mtod(mbuf,void *),mbuf->data_len);
+		total_len+=mbuf->data_len;
+		mbuf=mbuf->next;
+	}
+	return total_len;
+}
+
+static void pcap_dpdk_breakloop(pcap_t *p)
+{
+	pcap_breakloop_common(p);
+	struct pcap_dpdk *pd = (struct pcap_dpdk*)(p->priv);
+	pd->break_loop = p->break_loop;	
```

Comment:
So why does there need to be a private copy of the "break out of the loop" variable?

Reply:
OK, p->break_loop will be used directly.